src/System.IO.FileSystem/src/System/IO/PathHelpers.Windows.cs

```diff
@@ -2,80 +2,45 @@
-        internal static readonly char[] TrimEndChars = { (char)0x9, (char)0xA, (char)0xB, (char)0xC, (char)0xD, (char)0x20, (char)0x85, (char)0xA0 };
-        internal static readonly char[] TrimStartChars = { ' ' };
-
-        internal static bool ShouldReviseDirectoryPathToCurrent(string path)
-        {
-            // In situations where this method is invoked, "<DriveLetter>:" should be special-cased
-            // to instead go to the current directory.
-            return path.Length == 2 && path[1] == ':';
-        }
-
-        // ".." can only be used if it is specified as a part of a valid File/Directory name. We disallow
-        //  the user being able to use it to move up directories. Here are some examples eg
-        //    Valid: a..b  abc..d
-        //    Invalid: ..ab   ab..  ..   abc..d\abc..
-        //
-        internal static void CheckSearchPattern(string searchPattern)
+        internal static unsafe int GetRootLength(ReadOnlySpan<char> path)
         {
-            for (int index = 0; (index = searchPattern.IndexOf("..", index, StringComparison.Ordinal)) != -1; index += 2)
+            fixed (char* p = &path.DangerousGetPinnableReference())
             {
-                // Terminal ".." or "..\". File and directory names cannot end in "..".
-                if (index + 2 == searchPattern.Length ||
-                    PathInternal.IsDirectorySeparator(searchPattern[index + 2]))
-                {
-                    throw new ArgumentException(SR.Arg_InvalidSearchPattern, nameof(searchPattern));
-                }
+                return (int)GetRootLength(p, (uint)path.Length);
             }
         }
+    }
 
-        internal static string NormalizeSearchPattern(string searchPattern)
+    internal static partial class PathHelpers
+    {
+        internal static bool ShouldReviseDirectoryPathToCurrent(string path)
         {
-            Debug.Assert(searchPattern != null);
-
-            // Win32 normalization trims only U+0020.
-            string tempSearchPattern = searchPattern.TrimEnd(PathHelpers.TrimEndChars);
-
-            // Make this corner case more useful, like dir
-            if (tempSearchPattern.Equals("."))
-            {
-                tempSearchPattern = "*";
-            }
-
-            CheckSearchPattern(tempSearchPattern);
-            return tempSearchPattern;
+            // In situations where this method is invoked, "<DriveLetter>:" should be special-cased
+            // to instead go to the current directory.
+            return path != null && path.Length == 2 && path[1] == ':';
         }
 
-        internal static string GetFullSearchString(string fullPath, string searchPattern)
-        {
-            Debug.Assert(fullPath != null);
-            Debug.Assert(searchPattern != null);
+        internal static string TrimEndingDirectorySeparator(string path) =>
+            EndsInDirectorySeparator(path) ?
+                path.Substring(0, path.Length - 1) :
+                path;
```

Comment:
Could this work in terms of spans to avoid the substring?  Or the caller really needs a string?

Reply:
It can when we add the Span overloads to Path.